engine/player/player.cpp

```diff
@@ -3029,6 +3029,123 @@ void player_t::init_finished()
+    std::unordered_map<buff_t*, std::tuple<int, double, timespan_t>> precombat_buff_state;
+
+    for ( const auto& v : precombat_state_map )
+    {
+      auto splits = util::string_split<std::string>( v.first, "." );
+
+      if ( splits.size() < 2 )
+      {
+        sim->error( "Invalid 'override.precombat_state' option: '{}'", v.first );
+        continue;
+      }
+
+      auto type = splits[ 0 ];
+      auto name = splits[ 1 ];
+
+      if ( type == "buff" )
+      {
+        if ( splits.size() != 3 ){
+          sim->error( "Invalid 'override.precombat_state' buff expression: '{}'", v.first );
+          continue;
+        }
+
+        auto buff = buff_t::find( this, name );
+        if ( !buff )
+        {
+          sim->error( "No buff found for 'override.precombat_state' buff expression: '{}'", name );
+          continue;
+        }
+
+        bool new_buff_state = precombat_buff_state.find( buff ) == precombat_buff_state.end();
+
+        if ( new_buff_state )
+        {
+          // Initialize the values for a buff's stack, value, and duration
+          // parameters to the default parameter values of buff_t::execute.
+          precombat_buff_state[ buff ] = { -1, buff_t::DEFAULT_VALUE(), timespan_t::min() };
+        }
+
+        if ( splits[ 2 ] == "stack" )
+        {
+          std::get<0>( precombat_buff_state[ buff ] ) = util::to_int( v.second );
+          continue;
+        }
+        else if ( splits [ 2 ] == "value" )
+        {
+          std::get<1>( precombat_buff_state[ buff ] ) = util::to_double( v.second );
+          continue;
+        }
+        else if ( splits [ 2 ] == "remains" )
+        {
+          std::get<2>( precombat_buff_state[ buff ] ) = timespan_t::from_seconds( util::to_double( v.second ) );
+          continue;
+        }
+
+        // If no part of the newly created state was set, it needs to be
+        // removed to avoid triggering the buff with all default values.
+        if ( new_buff_state )
+        {
+          precombat_buff_state.erase( buff );
+        }
+
+        sim->error( "Invalid 'override.precombat_state' buff expression: '{}'", v.first );
+      }
+      else if ( type == "cooldown" )
+      {
+        if ( splits.size() != 2 ){
+          sim->error( "Invalid 'override.precombat_state' cooldown expression: '{}'", v.first );
+          continue;
+        }
+
+        auto cd = find_cooldown( name );
+        if ( !cd )
+        {
+          sim->error( "No cooldown found for 'override.precombat_state' cooldown expression: '{}'", name );
+          continue;
+        }
+
+        timespan_t duration = timespan_t::from_seconds( util::to_double( v.second ) );
+        add_precombat_cooldown_state( cd, duration );
+      }
+      else
+      {
+        sim->error( "Invalid type '{}' for 'override.precombat_state' option.", type );
```

Comment:
Might be better to throw here as well, if you are also doing that for the `!buff->allow_precombat` case further down. I would definitely treat this as invalid user input right away.